File: src/main/java/cc/mirukuneko/realtrainmodrenewed/network/ConfigureSpeakerPayload.java
package cc.mirukuneko.realtrainmodrenewed.network;

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed;
import cc.mirukuneko.realtrainmodrenewed.blockentity.InstalledObjectBlockEntity;
import cc.mirukuneko.realtrainmodrenewed.installedobject.SpeakerSoundConfig;
import io.netty.buffer.ByteBuf;
import net.minecraft.core.BlockPos;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import net.minecraft.server.level.ServerPlayer;
import net.neoforged.neoforge.network.PacketDistributor;
import net.neoforged.neoforge.network.handling.IPayloadContext;

import java.util.Arrays;

/**
 * クライアント → サーバー。 スピーカーGUIからの設定送信。
 * <ul>
 *   <li>{@code soundSlot >= 1}: 音源ID(1-64)に {@code soundName} を割り当て（全体共通・config保存）し、
 *       全クライアントへ同期する。</li>
 *   <li>{@code range >= 1}: そのスピーカーブロックの可聴範囲を設定する。</li>
 * </ul>
 * 変更しない項目は soundSlot=0 / range=0 を送る。
 */
public record ConfigureSpeakerPayload(BlockPos pos, int soundSlot, String soundName, int range)
        implements CustomPacketPayload {

    public static final Type<ConfigureSpeakerPayload> TYPE = new Type<>(
        Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "configure_speaker")
    );

    public static final StreamCodec<ByteBuf, ConfigureSpeakerPayload> STREAM_CODEC = StreamCodec.composite(
        BlockPos.STREAM_CODEC,
        ConfigureSpeakerPayload::pos,
        ByteBufCodecs.INT,
        ConfigureSpeakerPayload::soundSlot,
        ByteBufCodecs.STRING_UTF8,
        ConfigureSpeakerPayload::soundName,
        ByteBufCodecs.INT,
        ConfigureSpeakerPayload::range,
        ConfigureSpeakerPayload::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public static void handleOnServer(ConfigureSpeakerPayload payload, IPayloadContext context) {
        context.enqueueWork(() -> {
            if (!(context.player() instanceof ServerPlayer player)) {
                return;
            }
            if (payload.soundSlot() >= 1 && payload.soundSlot() <= SpeakerSoundConfig.MAX_SOUND_ID) {
                SpeakerSoundConfig.setSound(payload.soundSlot(), payload.soundName(), true);
                String[] snapshot = SpeakerSoundConfig.snapshot();
                SyncSpeakerSoundsPayload sync = new SyncSpeakerSoundsPayload(Arrays.asList(snapshot));
                if (player.level().getServer() != null) {
                    for (ServerPlayer p : player.level().getServer().getPlayerList().getPlayers()) {
                        PacketDistributor.sendToPlayer(p, sync);
                    }
                }
            }
            if (payload.range() >= 1
                && player.level().getBlockEntity(payload.pos()) instanceof InstalledObjectBlockEntity be
                && be.isSpeaker()) {
                be.setSpeakerRange(payload.range());
            }
        });
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/network/ConfigureSpeakerPayload.kt
package cc.mirukuneko.realtrainmodrenewed.network

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.blockentity.InstalledObjectBlockEntity
import cc.mirukuneko.realtrainmodrenewed.installedobject.SpeakerSoundConfig
import io.netty.buffer.ByteBuf
import net.minecraft.core.BlockPos
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.minecraft.server.level.ServerPlayer
import net.neoforged.neoforge.network.PacketDistributor
import net.neoforged.neoforge.network.handling.IPayloadContext
import java.util.Arrays

/** Client-to-server speaker GUI configuration payload. */
@JvmRecord
data class ConfigureSpeakerPayload(
    val pos: BlockPos,
    val soundSlot: Int,
    val soundName: String,
    val range: Int,
) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<ConfigureSpeakerPayload> = CustomPacketPayload.Type(
            Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "configure_speaker")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, ConfigureSpeakerPayload> = StreamCodec.composite(
            BlockPos.STREAM_CODEC,
            { payload -> payload.pos },
            ByteBufCodecs.INT,
            { payload -> payload.soundSlot },
            ByteBufCodecs.STRING_UTF8,
            { payload -> payload.soundName },
            ByteBufCodecs.INT,
            { payload -> payload.range },
            ::ConfigureSpeakerPayload,
        )

        @JvmStatic
        fun handleOnServer(payload: ConfigureSpeakerPayload, context: IPayloadContext) {
            context.enqueueWork {
                val player = context.player() as? ServerPlayer ?: return@enqueueWork
                if (payload.soundSlot in 1..SpeakerSoundConfig.MAX_SOUND_ID) {
                    SpeakerSoundConfig.setSound(payload.soundSlot, payload.soundName, true)
                    val sync = SyncSpeakerSoundsPayload(Arrays.asList(*SpeakerSoundConfig.snapshot()))
                    for (target in player.level().server.playerList.players) {
                        PacketDistributor.sendToPlayer(target, sync)
                    }
                }
                val blockEntity = player.level().getBlockEntity(payload.pos)
                if (payload.range >= 1 && blockEntity is InstalledObjectBlockEntity && blockEntity.isSpeaker) {
                    blockEntity.setSpeakerRange(payload.range)
                }
            }
        }
    }
}
